### ISSUE TITLE
[Skype for Business] HTML message formatting inconsistencies (macOS vs Windows)

### ISSUE THREAD

**Issue by amehta3000** [Microsoft/BotBuilder#3808]
## Bot Info 
* SDK Platform: Node.js
* SDK Version: BotBuilder 3.11.0
* Active Channels: Skype for Business
* Deployment Environment: ngrok

## Issue Description
Documentation indicates that we can send messages with **html** and it will format properly on the SFB channel. https://docs.microsoft.com/en-us/bot-framework/nodejs/bot-builder-nodejs-message-create

However, we are seeing dramatic inconsistencies from one platform to another using the same code.  Looking for guidance on how to develop and format our messages for Skype for Business so that macOS/iOS will also render html. 

We'd like to use the HTML format we are seeing in Skype for Business on windows. When the same code is rendered in macOS, it shows just raw html
 

## Code Example
Not setting textFormat, using the default "markdown".  
**Setting it to "xml" yields no noticeable change.**  
`response.text` is just raw text with html tags, looks similar to what is shown in macOS screenshot below.
```
var customMsg = new builder.Message(session).text(response.text);
session.send(customMsg);
```

## Expected Behavior
SKYPE for BUSINESS 
**(windows client)**
<img width="400" alt="sfb_html_markdownii 1" src="https://user-images.githubusercontent.com/151545/33097192-2d5de3e8-cebe-11e7-82b0-2b7038707c09.PNG">

## Actual Results
SKYPE for BUSINESS 
**(macOS client update 16.12.77)**
![image](https://user-images.githubusercontent.com/151545/33098567-aacab136-cec2-11e7-81d5-cb3dc3592b29.png)

**Comment by JasonSowers** [comment]
Thank you for reporting I will ping the Skype for business team.

**Comment by JasonSowers** [comment]
Still waiting on a response

**Comment by amehta3000** [comment]
Thx, anxiously waiting. Cheers

**Comment by amehta3000** [comment]
Hi there, any updates on this? This behavior is seriously impacting our customers who using Skype for Business on mac OS and iOS. Adding any sort of html added into our dialogs fails to render on iOS at all. Thanks.

**Comment by JasonSowers** [comment]
I still have no response, I will ping again.

**Comment by JasonSowers** [comment]
@amehta3000  from the sfb team

> Is it possible to get some mac client traces from customer to investigate deeper?

Are you able to provide this?

**Comment by amehta3000** [comment]
Hi, I'm hoping this is this client logs they are looking for outlined in this article: 
http://www.itprotoday.com/microsoft-exchange/mac-skype-business-2015-client-side-logs

@JasonSowers, can you please provide me an email address so I can send you the file privately. thx

**Comment by JasonSowers** [comment]
@amehta3000 v-jassow@microsoft.com

**Comment by amehta3000** [comment]
@JasonSowers sent

**Comment by JasonSowers** [comment]
I received them and have forwarded them to the sfb team, I will let you know as soon as I hear something.

**Comment by JasonSowers** [comment]
@amehta3000 can you prodive this?

From the sfb team:

> are you able to get customer’s bot input  in message Activity, (the http messsage that bot send to channel which contains that mark down part), I would like to do more experiment on this.

**Comment by amehta3000** [comment]
Sent privately to you via email, hope it suffices. LMK. thx

**Comment by JasonSowers** [comment]
There is an ongoing discussion I will update you as I get information.  At this point, I would look into possibly changing your HTML to markup.

**Comment by amehta3000** [comment]
Thanks Jason, 
I think it is interesting cause Markdown renders down to HTML so I would expect HTML to just work out the box.
That being said, I've started trying to use Markdown but it just doesn't render as well as HTML, and we lose the ability to show color along with formatting font sizes that I can get with HTML on a PC.  

Seem the only thing I can get is **bold** and _italics_ to work. 

My main frustration is that line breaks in Markdown don't seem to work as expected on either channel. Should be 2 spaces (\s\s) but both PC and Mac completely disregard it.  If I put <br> then PC will work but Mac will render the <br>.
I'll follow up with some additional screenshots. 

We appreciate your help and communication with the team there. thx for any more insight.

**Comment by amehta3000** [comment]
Hi Jason, just wondering if the final solution is to go straight markdown and assume that there is no formatting at all on a SFB mac client.  Markdown doesn't really render in Mac OS. It just end up being simple text.

Do you have any insight on how to issue a line break?  
We've been using  '\n\n\u200C<br\>' but honestly, that looks very hacky and not ideal results.  How do you programmatically get line breaks, when markdown says it is  '  \n' 
thx.

**Comment by JasonSowers** [comment]
@amehta3000  The  Only way I know a line break works for sure across all channels is using "\n\n"

**Comment by ganesh1190** [comment]
waiting for update , facing same issue , any Updates on it?

**Comment by amehta3000** [comment]
No updates. Was recommended to just go straight markdown, but as you may already know the SFB client on Mac doesn’t seem to even render markdown bold or italics either. Wondering if there will ever be an update or if there is an intention to make it better.

**Comment by ganesh1190** [comment]
Need only BOLD and ITALIC.
Any idea...? Please share.

**Comment by band888888** [comment]
Like  @ganesh1190  mention the same requirement on SFB.

 @amehta3000  I would like to consult you, "\n\n" it's working to break the link for the SFB on windows?
@JasonSowers I test it  "\n\n" it's working to break the link on the SFB on windows. Any idea?

**Comment by deepankar27** [comment]
Any update on this, markdown is not at all rendering to Mac OS devices. I am trying to use link tag but it's not working.

**Comment by bora89** [comment]
How can this issue be closed if problems is still there? MacOS - does not supported at all. And can you explain how we can use colors etc? As I am not sure if css is supported

**Comment by albigi** [comment]
we are receiving lots of inquires about what styling is supported on MacOS and what not.
is there some guideline we can point our customers at?

thanks!

**Comment by EricDahlvang** [comment]
Reopened.  Labeled as bug.

**Comment by brienmorton** [comment]
has this issue been resolved?

**Comment by EricDahlvang** [comment]
@brienmorton  No, this issue has not been resolved.  This issue is also not related to email, but to Skype for Business.

**Comment by brienmorton** [comment]
@EricDahlvang  thank you.   I have very large enterprise customer trying to deploy this.  Any idea when it might be resolved?  Or any work around?

**Comment by amehta3000** [comment]
From the contact we've had with MS SFB Team, there is no formatting supported, should be all done in markdown which also does not really do much. The html that works on PC via Bot Framework is also not recommended. It does not seem to be a priority to fix SFB as MS wants everyone to transition to MS Teams which is much more versatile. Granted this is not always an easy option with enterprises and legacy systems, so we have come to the conclusion to treat our bots on SFB more like SMS. Make your copy concise with single turn actions to keep it light and conversational. Throw in an emoji or 2 at best. Good luck.

**Comment by brienmorton** [comment]
Thank you both for your help.

**Comment by sgellock** [comment]
Thank you for opening an issue against the Bot Framework SDK v3. As part of the Bot Framework v4 release, we’ve moved all v3 work to a new repo located at https://github.com/microsoft/botbuilder-v3. We will continue to support and offer maintenance updates to v3 via this new repo.
 
From now on, https://github.com/microsoft/botbuilder repo will be used as hub, with pointers to all the different SDK languages, tools and samples repos.
 
As part of this restructuring, we are closing all tickets in this repo.
 
For defects or feature requests, please create a new issue in the new Bot Framework v3 repo found here:
https://github.com/microsoft/botbuilder-v3/issues

For Azure Bot Service Channel specific defects or feature requests (e.g. Facebook, Twilio, Teams, Slack, etc.), please create a new issue in the new Bot Framework Channel repo found here:
https://github.com/microsoft/botframework-services/issues
 
For product behavior, how-to, or general understanding questions, please use Stackoverflow.
https://stackoverflow.com/search?q=bot+framework


Thank you.
 
The Bot Framework Team
